src/content_script/input_tools.js

```diff
@@ -0,0 +1,237 @@
+let SLSettings = {};
+
+const InputTools = {
+  isLoading: false,
+
+  // store tracked input elements
+  trackedElements: [],
+
+  init(target) {
+    InputTools.queryEmailInputAndApply(target, (element) => {
+      if (!InputTools.isValidEmailInput(element)) {
+        return;
+      }
+
+      // ignore if this elements has already been tracked
+      const i = InputTools.trackedElements.indexOf(element);
+      if (i === -1) {
+        InputTools.trackedElements.push(element);
+        InputTools.addSLButtonToInput(element);
+      }
+    });
+  },
+
+  destroy(target) {
+    InputTools.queryEmailInputAndApply(target, (element) => {
+      // remove element from tracking list
+      const i = InputTools.trackedElements.indexOf(element);
+      if (i !== -1) {
+        InputTools.trackedElements.splice(i, 1);
+      }
+    });
+  },
+
+  queryEmailInputAndApply(target, actionFunction) {
+    if (!target.querySelectorAll) return;
+    const elements = target.querySelectorAll(
+      "input[type='email'],input[name*='email'],input[id*='email']"
+    );
+    for (const element of elements) {
+      actionFunction(element);
+    }
+  },
+
+  isValidEmailInput(element) {
+    const style = getComputedStyle(element);
+    return (
+      // check if element is visible
+      style.visibility !== "hidden" &&
+      style.display !== "none" &&
+      style.opacity !== "0" &&
+      // check if element is not disabled
+      !element.disabled &&
+      // for example, we must filter out a checkbox with name*=email
+      // check if element is text or email input
+      (element.type === "text" || element.type === "email")
+    );
+  },
+
+  addSLButtonToInput(inputElem, isInputFixed) {
+    // create wrapper for SL button
+    const btnWrapper = InputTools.newDiv("sl-button-wrapper");
+    const inputSumHeight = inputElem.getBoundingClientRect().height + "px";
+    btnWrapper.style.height = inputSumHeight;
+    btnWrapper.style.width = inputSumHeight;
+    document.body.appendChild(btnWrapper);
+
+    // create the SL button
+    const slButton = InputTools.newDiv("sl-button");
+    slButton.addEventListener("click", function () {
+      InputTools.handleOnClickSLButton(inputElem, slButton);
+    });
+    slButton.style.height = inputSumHeight;
+    slButton.style.width = inputSumHeight;
+    btnWrapper.appendChild(slButton);
+
+    InputTools.placeElementToTheRight(inputElem, btnWrapper, isInputFixed);
+  },
+
+  newDiv(classes) {
```

Comment:
should `classes` be renamed to `class`?

Reply:
No, because `class` is a reversed keyword in JS. However, it can be className

Reply:
fixed on cd4dd7e19e3a95cc8753ec48f2364eab409e66da